[SQL-gen] Generate queries for structs

c-harding opened the pull request:
Following from #372, with the TODO done: here are the SQL queries for interfacing with a struct in the database

c-harding committed:
[SQL-gen] Generate SQL queries for structs

c-harding committed:
[SQL-gen] Simplify conditionsToDisjunction

c-harding committed:
[SQL-gen] Use correct prepared symbols for language in SQL

c-harding committed:
Merge branch 'develop' into struct-root-queries

c-harding committed:
Merge remote-tracking branch 'origin/develop' into struct-root-queries

willjvsmith commented on src/test/scala/temple/builder/ServerBuilderTest.scala:
```diff
@@ -108,7 +108,10 @@ class ServerBuilderTest extends FlatSpec with Matchers {
           name = "Test",
           opQueries = SortedMap(
-            // TODO: Coming soon
+            Create -> "INSERT INTO test (favourite_colour, bed_time, favourite_number) VALUES ($1, $2, $3) RETURNING favourite_colour, bed_time, favourite_number;",
+            Read   -> "SELECT favourite_colour, bed_time, favourite_number FROM test WHERE id = $1;",
+            Update -> "UPDATE test SET favourite_colour = $1, bed_time = $2, favourite_number = $3 WHERE id = $4 RETURNING favourite_colour, bed_time, favourite_number;",
+            Delete -> "DELETE FROM test WHERE id = $1;",
```
I don't think we've included trailing semi-colons elsewhere

c-harding replied:
@smithwjv this is from mkCode.stmt, see the existing queries in https://github.com/TempleEight/temple/blob/develop/src/test/scala/temple/builder/ServerBuilderTest.scala

willjvsmith commented on src/main/scala/temple/builder/ServerBuilder.scala:
```diff
@@ -27,34 +27,34 @@ object ServerBuilder {
   ): ServiceRoot = {
-    val language = serviceBlock.lookupMetadata[ServiceLanguage].getOrElse(ProjectConfig.defaultLanguage)
-    val database = serviceBlock.lookupMetadata[Database].getOrElse(ProjectConfig.defaultDatabase)
+    val language = serviceBlock.lookupMetadata[ServiceLanguage] getOrElse ProjectConfig.defaultLanguage
+    val database = serviceBlock.lookupMetadata[Database] getOrElse ProjectConfig.defaultDatabase
```
why?

c-harding replied:
I was moving the two lines from below (`readable`, `writable`) up, and one of them was infix already, so I made them consistent. Happy to switch them all to methods if you’d prefer

willjvsmith replied:
I think we've avoided this notation elsewhere that isn't tests

c-harding replied:
fixed

jaylees14 commented on src/main/scala/temple/builder/ServerBuilder.scala:
```diff
@@ -27,34 +27,34 @@ object ServerBuilder {
 
-    // The created by attribute exists if the project has auth and this service does not have an auth block
-    val createdBy =
-      if (projectUsesAuth && !hasAuthBlock)
```
Do we not need this part of the condition?

c-harding replied:
@jaylees14 this is calculated when we add a CreatedByAttribute, so it’s already been done (and the condition is slightly more complex now: struct blocks don’t have createdBy)

willjvsmith commented on src/main/scala/temple/builder/ServerBuilder.scala:
```diff
@@ -27,34 +27,34 @@ object ServerBuilder {
     // Whether or not this service has an auth block
     val hasAuthBlock = serviceBlock.lookupLocalMetadata[ServiceAuth].isDefined
 
-    // The created by attribute exists if the project has auth and this service does not have an auth block
-    val createdBy =
-      if (projectUsesAuth && !hasAuthBlock)
-        Some(CreatedByAttribute(languageConfig.createdByInputName, languageConfig.createdByName))
-      else None
+    val createdBy = when(serviceBlock.attributes.valuesIterator contains AbstractAttribute.CreatedByAttribute) {
+      CreatedByAttribute(languageConfig.createdByInputName, languageConfig.createdByName)
+    }
```
Is this equivalent?

c-harding replied:
@smithwjv `when` is equivalent to the if-then-some-else-none. The condition has now been moved to the [validator](https://github.com/TempleEight/temple/blob/730a8c27462c98346f766e73cbfb9b4c176609a0/src/main/scala/temple/DSL/semantics/Validator.scala#L37), so it’s calculated in just one place rather than both, but otherwise the same

willjvsmith replied:
Yeah I was referring to the condition, I see, fair enough

willjvsmith reviewed (approved):
LGTM 🚀 

You've done some refactoring here that isn't mentioned in the description

jaylees14 commented on src/test/scala/temple/builder/DatabaseBuilderTest.scala:
```diff
@@ -90,9 +95,12 @@ class DatabaseBuilderTest extends FlatSpec with Matchers {
     queries(CRUD.List) shouldBe DatabaseBuilderTestData.sampleListStatementEnumerateByCreator
   }
+
+  // TODO: test the list endpoint with isStruct=true
```
This would actually be quite good to do since it tests the disjunction logic. Could you add one please

c-harding replied:
I’ve realised this will never happen anyway, and so removed that disjunction function: a struct can never have createdBy, so it’s one or the other, but not both

c-harding committed:
[SQL-gen] Remove conditionsToDisjunction

Structs don’t have a createdBy attribute, so listing can never have both conditions

c-harding committed:
[SQL-gen] Remove infix of Option#getOrElse
